fedn/fedn/network/controller/control.py

```diff
@@ -98,118 +114,126 @@ def session(self, config):
-            self.tracer.set_round_data(round_data)
-
         # TODO: Report completion of session
         self._state = ReducerState.idle
 
     def round(self, session_config, round_id):
-        """ Execute a single global round.
+        """ Execute one global round.
 
         :param session_config: The session config.
         :type session_config: dict
         :param round_id: The round id.
-        :type round_id: str(int)
+        :type round_id: str
+
         """
 
-        round_data = {'round_id': round_id}
+        self.new_round({'round_id': round_id, 'status': "Pending"})
 
         if len(self.network.get_combiners()) < 1:
-            print("REDUCER: No combiners connected!", flush=True)
-            round_data['status'] = 'Failed'
-            return None, round_data
+            print("CONTROLLER: Round cannot start, no combiners connected!", flush=True)
+            self.set_round_status(round_id, 'Failed')
+            return None, self.statestore.get_round(round_id)
 
-        # 1. Assemble round config for this global round,
-        # and check which combiners are able to participate
-        # in the round.
+        # Assemble round config for this global round
         round_config = copy.deepcopy(session_config)
         round_config['rounds'] = 1
         round_config['round_id'] = round_id
         round_config['task'] = 'training'
         round_config['model_id'] = self.statestore.get_latest_model()
         round_config['helper_type'] = self.statestore.get_helper()
 
-        combiners = self.get_participating_combiners(round_config)
-        round_start = self.evaluate_round_start_policy(combiners)
+        self.set_round_config(round_id, round_config)
+
+        # Get combiners that are able to participate in round, given round_config
+        participating_combiners = self.get_participating_combiners(round_config)
+
+        # Check if the policy to start the round is met
+        round_start = self.evaluate_round_start_policy(participating_combiners)
 
         if round_start:
-            print("CONTROL: round start policy met, participating combiners {}".format(
-                combiners), flush=True)
+            print("CONTROL: round start policy met, {} participating combiners.".format(
+                len(participating_combiners)), flush=True)
         else:
             print("CONTROL: Round start policy not met, skipping round!", flush=True)
-            round_data['status'] = 'Failed'
-            return None
+            self.set_round_status(round_id, 'Failed')
+            return None, self.statestore.get_round(round_id)
+
+        # Ask participating combiners to coordinate model updates
+        _ = self.request_model_updates(participating_combiners)
+        # TODO: Check response
 
-        round_data['round_config'] = round_config
+        # Wait until participating combiners have produced an updated global model,
+        # or round times out.
+        def do_if_round_times_out(result):
```

Comment:
I'm not a fan of inner/nested functions, bad for readability and "round" is complex already

Reply:
I am not sure if there is another way to do it if we want to use the retrying library.